thunder/benchmarks/benchmark_litgpt.py

```diff
@@ -407,7 +425,8 @@ def train(self):
                     input_ids, targets = next(self.train_data_iter)
                     input_ids = input_ids.to(self.device)
                     targets = targets.to(self.device)
-                    logits = self.model(input_ids)
+                    with te.fp8_autocast(enabled=is_transformer_engine(self.low_precision_mode)):
```

Comment:
Should this also depend on whether we're running a thunder benchmark (similar to the `__init__`)? (also for the context manager below)

Reply:
I agree, let's create separate codepaths for with vs without TE.

Reply:
Hello @t-vi, could you explain a little bit more? You mean that for thunder case, we do not need to provide a context manager?

Reply:
Yes, I have been wondering about the use of the context manager for thunder, and in particular the setup being conditioned to
```
# fp8-delayed-te precision for thunder is handled through compile
if is_transformer_engine(low_precision_mode) and "thunder" not in self.compile:
   ...
```

and then the context manager uses
```
with te.fp8_autocast(enabled=is_transformer_engine(self.low_precision_mode)):
```

We could have
- add `and "thunder" not in self.compile` to the enabled argument,
- a path without the context manager (or a `contextlib.nullcontext` manager) when not using `te.fp8_autocast`.

What happens when we hit the current context manager and we did not convert the module? To my mind, we need to be sure to have something that works in all four settings (thunder + no fp8, thunder + fp8, no thunder + no fp8, no thunder + fp8), and I'm not entirely sure that the current PR covers them.

Reply:
I see. I thought it is also necessary to provide a context manager for thunder - I will make the required changes to straighten it out.

For the H100/GH200 remark, that is true. FP8-enabled script will throw error if we run for example for A100. I am not sure how to handle this - revert to pure BF16 run, or maybe throw an explicit error why it fails?

Reply:
> maybe throw an explicit error why it fails?

Yes, please! In the benchmarking context it is best to get error feedback when a requested config does not work.

Reply:
When using fp8 with `thunder`, `thunder.jit` takes care of wrapping the jitted function with `te.fp8_autocast`. Even, if the user runs this jitted function under their `te.fp8_autocast`, computation would still adhere to `thunder`'s context as it is the inner context. So, there is no need of `te.fp8_autocast` when using `thunder + transformer_engine_ex`.

See this [PR](https://github.com/Lightning-AI/lightning-thunder/pull/48) which added the support.

-----

For other cases, `te.fp8_autocast(enabled=True)` should have no effect when there are no TE layers in the model but it would be confusing to see this for a non-TE model. So, we should only enable it when TE mode is active.

One way to achieve this would be to optional switch context between `contextlib.nullcontext` or `te.fp8_autocast` depending on `is_transformer_engine` as suggested by @t-vi 

Other approach could be to replace `self.model` with a function which will call the model under `fp8_autocast` context (see example below). I like as there would be fewer places where you will have to condition on `is_transformer_engine` and fewer places to forget adding this context.

```python
if is_transformer_engine(low_precision_mode) and "thunder" not in self.compile:
    te_precision = TransformerEnginePrecision(weights_dtype=torch.bfloat16, replace_layers=True)
    converted_model = te_precision.convert_module(self.model)
    def call_model_with_fp8autocast(*args, **kwargs):
         with te.fp8_autocast(enabled=True):
             return converted_model(*args, **kwargs)
    self.model = call_model_with_fp8autocast
```
-------

Also, another thing to note regarding benchmark is that `TransformerEnginePrecision` will replace `Linear` and `LayerNorm` with their TE equivalent while `thunder.jit` with `transformer_engine_ex` currently only replaces the `Linear`. So, the models won't be similar in terms of the layers which do their computation in FP8.


https://lightning.ai/docs/fabric/stable/api/generated/lightning.fabric.plugins.precision.TransformerEnginePrecision.html

Reply:
@kshitij12345 @t-vi @lantiga I applied changes according to your comments - hopefully I did not misunderstand your comments. Let me know if something else is needed or requires modification :)

Reply:
I'm in favor of a completely separate path when TE is not used (`nullcontext` is useful to avoid code duplication but in this case the impact is minimal at the cost of clarity).

Also let's make importing TE optional at the top of the script, so we don't strictly require it when we run benchmarks. You can try-catch the import and set a boolean variable accordingly.

Reply:
@lantiga What do you mean by "completely separate path"? Do you mean creating dedicated `def train(self)` method, sth like `def train_fp8(self):`? Or creating a separate module for fp8 training (`benchmark_litgpt_fp8.py`)?